### PR TITLE
[2.0] Make all inner methods and properties protected

### PR DESCRIPTION
**Pull request by nokitakaze**
This pull request **breaks** back compatibility.

### CODE REVIEW

File: tests/Processor/SanitizeDataProcessorTest.php [hunk 1]
```diff
@@ -157,7 +159,9 @@ public function testOverridenSanitize($processorOptions, $client_options, $dsn)
         /**
          * @var SanitizeDataProcessor $processor
          */
-        $processor = $client->processors[0];
+        $processors = new \ReflectionProperty($client, 'processors');
```
Comment (ste93cry) [inline tests/Processor/SanitizeDataProcessorTest.php]: PHPUnit provides an `assertAttributeEquals` method to check whether an attribute (even private) equals a given value

File: tests/Processor/SanitizeStacktraceProcessorTest.php [hunk 2]
```diff
@@ -65,9 +68,12 @@ public function testProcessWithPreviousException()
             $this->client->captureException($exception);
         }
 
-        $this->processor->process($this->client->_pending_events[0]);
+        $_pending_events = new \ReflectionProperty($this->client, '_pending_events');
```
Comment (ste93cry) [inline tests/Processor/SanitizeStacktraceProcessorTest.php]: PHPUnit provides an `assertAttributeEquals` method to check whether an attribute (even private) equals a given value

File: tests/Processor/SanitizeStacktraceProcessorTest.php [hunk 3]
```diff
@@ -42,9 +42,12 @@ public function testProcess()
...
```
Comment (ste93cry) [inline tests/Processor/SanitizeStacktraceProcessorTest.php]: PHPUnit provides an `assertAttributeEquals` method to check whether an attribute (even private) equals a given value

File: lib/Raven/Stacktrace.php [hunk 4]
```diff
@@ -171,7 +171,7 @@ public function addFrame($file, $line, array $backtraceFrame)
      *
      * @throws \OutOfBoundsException If the index is out of range
      */
-    public function removeFrame($index)
+    protected function removeFrame($index)
```
Comment (ste93cry) [inline lib/Raven/Stacktrace.php]: This should not be protected, an user may want to remove a given frame from the stacktrace before sending it to the Sentry server

File: tests/ClientTest.php [hunk 5]
```diff
@@ -334,10 +344,18 @@ public function testDsnFirstArgument()
     {
         $client = new Dummy_Raven_Client('http://public:secret@example.com/1');
 
-        $this->assertEquals(1, $client->project);
-        $this->assertEquals('http://example.com/api/1/store/', $client->server);
-        $this->assertEquals('public', $client->public_key);
-        $this->assertEquals('secret', $client->secret_key);
+        $project = new \ReflectionProperty($client, 'project');
```
Comment (ste93cry) [inline tests/ClientTest.php]: PHPUnit provides an `assertAttributeEquals` method to check whether an attribute (even private) equals a given value

File: tests/TransactionStackTest.php [hunk 6]
```diff
@@ -30,7 +30,9 @@ public function testSimple()
         $this->assertEquals($stack->peek(), null);
 
         $stack->clear();
-        $this->assertInternalType('array', $stack->stack);
-        $this->assertEquals(0, count($stack->stack));
+        $stack_r = new \ReflectionProperty($stack, 'stack');
```
Comment (ste93cry) [inline tests/TransactionStackTest.php]: PHPUnit provides an `assertAttributeEquals` method to check whether an attribute (even private) equals a given value

File: tests/StacktraceTest.php [hunk 7]
```diff
@@ -198,7 +198,9 @@ public function testRemoveFrame($index, $throwException)
 
         $this->assertCount(2, $stacktrace->getFrames());
 
-        $stacktrace->removeFrame($index);
+        $removeFrame = new \ReflectionMethod($stacktrace, 'removeFrame');
```
Comment (ste93cry) [inline tests/StacktraceTest.php]: This method should be public, so there is no need to use reflection here

File: tests/Processor/SanitizeStacktraceProcessorTest.php (hunk 3 above)
Comment (nokitakaze) [inline tests/Processor/SanitizeStacktraceProcessorTest.php]: I do not assert this value. I am doing magic with this value

Comment (nokitakaze) [inline tests/Processor/SanitizeStacktraceProcessorTest.php]: I do not assert this value. I am doing magic with this value

**Comment by nokitakaze** [comment]
.
.
.
uopz can not be compiled with PHP 7.1 at this state.
```
make: *** [src/class.lo] Error 1
ERROR: `make' failed
```

It is beautiful

File: tests/Processor/SanitizeStacktraceProcessorTest.php (hunk 2 above)
Comment (nokitakaze) [inline tests/Processor/SanitizeStacktraceProcessorTest.php]: Same here. I do not assert this value

**Comment by ste93cry** [comment]
I don't know what `uopz` extension does, but I never heard of it. Maybe we can remove it?

**Comment by Jean85** [comment]
Since we are breaking BC, why not make something private? Or do we expect end users to extend and be able to touch everything?

**Comment by tobias-nitsche** [comment]
@Jean85 my opinion to that:

protected functions are nice
protected properties are a mess

**Comment by nokitakaze** [comment]
Private methods and properties are existed only for prohibition of extending class in the one way or another. You SHOULD NOT use "private" until you want to prohibit override some behaviour. Word "final" id used for the same reason.

Look at this 33b9a36 / #409

**Comment by Jean85** [comment]
@nokitakaze I do not agree. Private does not prohibit extension (that's what `final` is for, as you are saying), it just reduces visibility.

If a method or a property is not needed outside of the current class, it should be private; it always can be upgraded to protected later. Having something non-private means that is a part of the contract with the end user, so it cannot be changed without a BC. So IMHO reducing the size of the contract is in our interest, to have less constraints on future developments.

**Comment by nokitakaze** [comment]
> is not needed outside

You don\'t know needed this property/method or not until you write extending class.

**Comment by Jean85** [comment]
That can be a valid reason. But this means that we are taking the conscious decision of willingly support the BC for every property and method.

I'm fine with that, but it should be agreed upon.

**Comment by tobias-nitsche** [comment]
@nokitakaze I disagree as well. You actually can modify properties of the parent class via set-Methods. Using this pattern, gives the parent class the possibiliy to control the input data. 

Without these protection, and can be hard to follow and observ changes of properties.

**Comment by nokitakaze** [comment]
@tobias-nitsche this should be using only if you sanitize all input data in your setters

**Comment by tobias-nitsche** [comment]
@nokitakaze To make it clear, protected functions are violating the open/closed principle https://en.wikipedia.org/wiki/Open/closed_principle

Maybe a sanitization is not yet needed, but changing variables directly in a child class, makes it hard to add this possibily in future. But it's not my decision.

**Comment by nokitakaze** [comment]
@tobias-nitsche , We override methods because we want to override/modificate a behaviour. If we don\'t broke code contract in children we don\'t break [LSP](https://en.wikipedia.org/wiki/Liskov_substitution_principle). So there is no any bad aftermath.

All we need to do is not to break code contract while overriding methods.

**Comment by tobias-nitsche** [comment]
Sorry, my bad. I meant protected properties. Not functions. (I corrected it)

**Comment by Jean85** [comment]
This should be completely redone on the top of 2.0 as it is now.

**Comment by nokitakaze** [comment]
Or just closed
